#333: aggregated reports

MihaStemberger opened the pull request:
Hello, here is my shot at trying to solve https://github.com/mojohaus/versions-maven-plugin/issues/333

ajarmoniuk commented on versions-common/src/main/java/org/codehaus/mojo/versions/utils/PropertyComparator.java:
```diff
@@ -42,14 +41,18 @@ public enum PropertyComparator implements Comparator<Property>
+                        : p2 == null
+                                ? -1
+                                : ( r = StringUtils.compare( p1.getName(), p2.getName() ) ) == 0
+                                        ? StringUtils.compare( p1.getValue(), p2.getValue() )
```
The original idea behind omitting the value in comparison was that properties with different values were treated as the same property. Was there a reason you wanted to treat properties with different values as different properties?

MihaStemberger replied:
The reason is that a child may define its own property value. If comparison is done only by name, then only one property entry will be in the final report and not all, that are possible.

ajarmoniuk commented on versions-common/src/main/java/org/codehaus/mojo/versions/api/Property.java:
```diff
@@ -176,4 +184,14 @@ public void setBanSnapshots( boolean banSnapshots )
         this.banSnapshots = banSnapshots;
     }
+
+    public String getValue()
```
The original idea behind omitting the value in comparison was that properties with different values were treated as the same property. That's why the Property class didn't have a value property.

MihaStemberger replied:
[See above](#discussion_r1038304268-permalink)

MihaStemberger commented on versions-maven-plugin/src/it/it-dependency-updates-aggregate-report-issue-333/src/site/site.xml:
```diff
@@ -0,0 +1,7 @@
+    <artifactId>maven-default-skin</artifactId>
+    <version>1.0</version>
+  </skin>
+</project>
```
Added

slawekjaranowski commented on versions-maven-plugin/src/it-repo/dummy-aggregate-child-333.pom:
```diff
@@ -0,0 +1,51 @@
+  <packaging>pom</packaging>
+
+  <properties>
+    <version.maven-clean-plugin>2.1</version.maven-clean-plugin>
+    <version.maven-deploy-plugin>2.1</version.maven-deploy-plugin>
```
Can we use a little newer versions in test or a dummy dependency

MihaStemberger replied:
Using newer version of `maven-clean-plugin`, also revisited and changed test poms, so they mainly use dummy dependencies.

slawekjaranowski commented:
For new files we use standard Apache 2.0 license headers: 
https://groups.google.com/g/mojohaus-dev/c/OTg7XfiX61w

I prefer groovy scripts for verification than beanshell ... as nice to have 😄

MihaStemberger commented on versions-maven-plugin/src/it/it-plugin-updates-aggregate-report-issue-333/pom.xml:
```diff
@@ -0,0 +1,75 @@
+        <reportSets>
+          <reportSet/>
+        </reportSets>
+      </plugin>
```
Added the configuration to `dummy-aggregate-parent-333.pom` and removed `maven-project-info-reports-plugin` from all child poms

MihaStemberger commented:
> For new files we use standard Apache 2.0 license headers: https://groups.google.com/g/mojohaus-dev/c/OTg7XfiX61w
> 
> I prefer groovy scripts for verification than beanshell ... as nice to have smile

Added `Apache 2.0 license headers` to newly added *.java files, replaced beanshell verification scripts with groovy scripts

slawekjaranowski commented on versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/AbstractDependencyUpdatesReportMojo.java:
```diff
@@ -0,0 +1,368 @@
+/**
+ * Generates a report of available updates for the dependencies of a project.
+ *
+ * @author Stephen Connolly
+ * @since 1.0-beta-1
```
copy paste 😄

slawekjaranowski commented on versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/AbstractPluginUpdatesReportMojo.java:
```diff
@@ -0,0 +1,252 @@
+/**
+ * Generates a report of available updates for the plugins of a project.
+ *
+ * @author Stephen Connolly
+ * @since 1.0-beta-1
```
copy paste 😄

slawekjaranowski commented on versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/AbstractPropertyUpdatesReportMojo.java:
```diff
@@ -0,0 +1,221 @@
+ * Generates a report of available updates for properties of a project which are linked to the dependencies and/or
+ * plugins of a project.
+ *
+ * @author Stephen Connolly
+ * @since 1.0-beta-1
```
copy paste 😄

slawekjaranowski commented on versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/PluginUpdatesAggregateReportMojo.java:
```diff
@@ -0,0 +1,91 @@
+        super( i18n, repositorySystem, aetherRepositorySystem, wagonMap, rendererFactory );
+    }
+
+    /**
+     * {@inheritDoc}
+     * */
```
not need IMHO

slawekjaranowski reviewed (commented):
last nits

MihaStemberger commented:
> last nits

Resolved

ajarmoniuk commented:
Yes, you're supposed to have edited https://github.com/mojohaus/versions-maven-plugin/blob/master/versions-maven-plugin/src/site/markdown/index.md#reporting-goals-overview 

But I guess you can still do it in another PR ;)
